csharp/src/Ice/InputStream-Class.cs

```diff
@@ -32,51 +32,42 @@ public void IceEndSlice()
-        // typeId is the expected type ID of this slice.
-        // firstSlice is true when reading the first (most derived) slice of an instance.
+        // typeId is the expected typeId of this slice.
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public void IceStartSlice(string typeId, bool firstSlice)
+        public void IceStartFirstSlice(string typeId, AnyClass v, bool setSlicedData = false)
         {
             Debug.Assert(InEncapsulation);
-            if (firstSlice)
-            {
-                Debug.Assert(_current != null && (_current.SliceTypeId == null || _current.SliceTypeId == typeId));
-                if (_current.InstanceType == InstanceType.Class)
-                {
-                    // For exceptions, we read it for the first slice in ThrowException.
-                    ReadIndirectionTableIntoCurrent();
-                }
+            Debug.Assert(_current != null && (_current.SliceTypeId == null || _current.SliceTypeId == typeId));
+            Debug.Assert(_current.InstanceType == InstanceType.Class);
 
-                // We can discard all the unknown slices: the generated code calls IceStartSliceAndGetSlicedData to
-                // preserve them and it just called IceStartSlice instead.
-                _current.Slices = null;
-            }
-            else
-            {
-                string? headerTypeId = ReadSliceHeaderIntoCurrent();
-                Debug.Assert(headerTypeId == null || headerTypeId == typeId);
-                ReadIndirectionTableIntoCurrent();
-            }
+            FirstSliceInit(v, setSlicedData);
+            ReadIndirectionTableIntoCurrent();
         }
 
-        // Starts reading the first slice of an instance and get the unknown slices for this instances that were
-        // previously saved (if any).
+        // Starts reading the first slice of a remote exception.
         // This is an Ice-internal method marked public because it's called by the generated code.
         // typeId is the expected typeId of this slice.
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public SlicedData? IceStartSliceAndGetSlicedData(string typeId)
+        public SlicedData? IceStartFirstSlice(string typeId)
         {
             Debug.Assert(InEncapsulation);
-            // Called by generated code for first slice instead of IceStartSlice
-            Debug.Assert(_current != null && (_current.SliceTypeId == null || _current.SliceTypeId == typeId));
-            if (_current.InstanceType == InstanceType.Class)
-            {
-                    // For exceptions, we read it for the first slice in ThrowException.
-                    ReadIndirectionTableIntoCurrent();
-            }
-            return SlicedData;
+            Debug.Assert(_current != null && _current.SliceTypeId == typeId);
+            Debug.Assert(_current.InstanceType == InstanceType.Exception);
+            return GetSlicedData();
```

Comment:
This would require passing the exception instance to the InputStream. We pass the AnyClass instance for classes primarily because we need to insert this instance in the instance map. For exceptions, it would be only to set this `SlicedData`, and this would be a somewhat odd way to set this field.